posit-dev/positron
# Add a restart command and button for Positron-backed notebooks

### Intent

> Describe briefly what problem this pull request resolves, or what new feature it introduces. Include screenshots of any new or altered UI. Link to any GitHub issues that are related.

Address #2213.

### Approach

> Describe the approach taken and the tradeoffs involved if non-obvious; add an overview of the solution if it's complicated.

Added a `NotebookSessionService.restartRuntimeSession` method very similar to the existing `{start,shutdown}RuntimeSession` methods.

Note that most line changes here are because I also refactored how we handle these operations: instead of using a deferred promise, I extracted a separate method in each case that handles the operation but doesn't fiddle with the map of promises tracking the operation, then wrapped that promise in the public method. This way, we can specify what happens on resolve/reject of the inner method in one place, and it feels more readable to me.

We only show our restart button the menu if a `positron-notebook-controller` kernel is used, and we disable the command (which greys out the button) if there is no session for the active notebook. The latter is tracked by the new `positron.hasRunningNotebookSession` context key.

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues.

There should be a restart button, and it should play well with opening/closing the notebook or executing a cell, in any order and even in quick succession.
